plat/drivers/uart/ns16550.c

```diff
@@ -65,26 +63,64 @@ static uint64_t ns16550_uart_base = CONFIG_EARLY_UART_NS16550_BASE;
 #else
 static uint8_t ns16550_uart_initialized;
 static uint64_t ns16550_uart_base;
+static uint32_t ns16550_reg_shift = 0; /* device-tree spec v0.4 Sect. 4.2.2 */
+static uint32_t ns16550_reg_width = 1; /* 8-bit register width by default */
```

Comment:
Good point, thanks! 👍🏼 @zhxj9823 will you please update to add these?

Reply:
> I have added the config options. Please check if these options work fine.

I think these two options should not be limited to `EARLY_UART`, but should be overriding the defaults when set, unconditionally. That would allow the driver to work correctly in any future platforms that do not use dts.

The priority should therefore be:
1. dts (overrides everything)
2. Kconfig (if set override defaults)
3. Defaults (according to the spec)

> It also raises the question of how to switch between pl011 and ns16550 as the early debug console.

For that we should merge the options of the different drivers into a single one, and most likely also consider an option in the command line, but this is outside the scope of this PR.

Reply:
I had a brief look on how the reg-shift and width for the `earlycon` are specified on linux. There these are not compiled in, but rather derived from the command-line options:
* https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html
* https://elixir.bootlin.com/linux/latest/source/drivers/tty/serial/earlycon.c#L93

The latter derives the regshift from the mmio width, ie 0 by default, 1 for MMIO16, and 2 for MMIO32 access. Right now we use 16-bit access and a regshift of 2 which doesn't make sense. @kubanrob should that be 32-bit access instead? Also would the command line be a valid option for earlycon in your case, assuming we implemented that? I imagine that since it's used on linux, it should be no problem for Unikraft.